pkg/kubelet/certificate/kubelet.go

```diff
@@ -90,13 +92,28 @@ func NewKubeletServerCertificateManager(kubeClient clientset.Interface, kubeCfg
 			StabilityLevel: compbasemetrics.ALPHA,
 		},
 	)
-	legacyregistry.MustRegister(certificateRotationAge)
+	registerMetricsOnce.Do(func() {
+		legacyregistry.MustRegister(certificateRenewFailure)
+		legacyregistry.MustRegister(certificateRotationAge)
+	})
 
 	getTemplate := func() *x509.CertificateRequest {
 		hostnames, ips := addressesToHostnamesAndIPs(getAddresses())
-		// don't return a template if we have no addresses to request for
-		if len(hostnames) == 0 && len(ips) == 0 {
-			return nil
+		// don't return a template if we have no IP addresses or DNS names to request for
```

Comment:
I don't know why, but I'm finding the combination of the nesting and the feature gate name really hard to follow. Would something like this be clearer?

```go
// by default, require at least one IP before requesting a serving certificate
hasRequiredAddresses := len(ips) > 0

// optionally allow requesting a serving certificate with just a DNS name
if utilfeature.DefaultFeatureGate.Enabled(features.AllowDNSOnlyNodeCSR) {
  hasRequiredAddresses = hasRequiredAddresses || len(hostnames) > 0
}

// don't return a template if we have no addresses to request for
if !hasRequiredAddresses {
  return nil
}
```

Reply:
is hard to follow but cleaner when we remove the feature gate, the diff is simpler

Reply:
is hard to follow but it is cleaner, so when we remove the feature gate, the diff is simpler

Reply:
will update with your suggestion